File: runtime/c/effects.c
#include "effects.h"
#include "store.h"
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <time.h>

static bool json_safe(const char *text) {
    size_t index;
    if (text == NULL) return true;
    for (index = 0u; text[index] != '\0'; ++index) if ((unsigned char)text[index] < 0x20u || text[index] == '\\' || text[index] == '"') return false;
    return true;
}

static LanaError store_attempt(LanaStore *store, LanaVM *vm,
                               const LanaEffectAttempt *attempt, bool allow_overwrite) {
    char key[64], json[2048]; Value existing, value; LanaError error;
    if (store == NULL || vm == NULL || attempt == NULL || attempt->struct_size < sizeof(*attempt) ||
        attempt->schema_version != 1u || attempt->attempt_id == 0u || attempt->decision_id == 0u ||
        attempt->effect_desc == NULL || attempt->failure_reason == NULL || attempt->status > EFFECT_STATUS_CANCELLED)
        return LANA_ERR_SCHEMA;
    if (!json_safe(attempt->effect_desc) || !json_safe(attempt->target) || !json_safe(attempt->failure_reason))
        return LANA_ERR_SCHEMA;
    (void)snprintf(key, sizeof(key), "effect-attempt/%llu/%llu",
                   (unsigned long long)attempt->decision_id,
                   (unsigned long long)attempt->attempt_id);
    if (!allow_overwrite) {
        error = lana_store_get(store, vm, key, &existing);
        if (error == LANA_OK) return LANA_ERR_CONFLICT;
        if (error != LANA_ERR_NOT_FOUND) return error;
    }
    (void)snprintf(json, sizeof(json),
                   "{\"attempt_id\":%llu,\"decision_id\":%llu,\"effect\":\"%s\",\"target\":\"%s\",\"end_time\":%llu,\"failure_reason\":\"%s\",\"schema\":1,\"start_time\":%llu,\"status\":%u}",
                   (unsigned long long)attempt->attempt_id, (unsigned long long)attempt->decision_id,
                   attempt->effect_desc, attempt->target == NULL ? "" : attempt->target,
                   (unsigned long long)attempt->end_time, attempt->failure_reason,
                   (unsigned long long)attempt->start_time, (unsigned)attempt->status);
    if (strlen(json) >= sizeof(json) - 1u) return LANA_ERR_LIMIT;
    value = lana_value_string(json); return lana_store_put(store, key, value);
}

LanaError lana_effect_execute_planned(LanaVM *vm, const Value *plan,
                                      LanaEffectExecutor executor, void *context,
                                      Value *out) {
    if (vm == NULL || plan == NULL || executor == NULL || out == NULL)
        return LANA_ERR_INVALID_STATE;
    return lana_vm_execute_planned_effect(vm, plan, executor, context, out);
}

LanaError lana_effect_store_attempt(LanaStore *store, LanaVM *vm,
                                    const LanaEffectAttempt *attempt) {
    return store_attempt(store, vm, attempt, false);
}

LanaError lana_effect_execute_attempt(LanaStore *store, LanaVM *vm,
                                      const LanaDecision *decision, uint64_t attempt_id,
                                      const Value *plan, LanaEffectExecutor executor,
                                      void *context, Value *out,
                                      LanaEffectAttempt *out_attempt) {
    LanaEffectAttempt attempt = {0}, pending = {0}; time_t started, ended; LanaError error, execution_error;
    if (decision == NULL || decision->outcome != POLICY_OUTCOME_AUTHORIZE || decision->effect == NULL || attempt_id == 0u)
        return LANA_ERR_CAPABILITY;
    started = time(NULL); if (started < 0) return LANA_ERR_IO;
    pending.struct_size = sizeof(pending); pending.schema_version = 1u; pending.attempt_id = attempt_id;
    pending.decision_id = decision->decision_id; pending.effect_desc = decision->effect;
    pending.target = decision->target;
    pending.status = EFFECT_STATUS_PENDING; pending.start_time = (uint64_t)started; pending.failure_reason = "";
    if ((error = store_attempt(store, vm, &pending, false)) != LANA_OK ||
        (error = lana_store_commit(store, NULL)) != LANA_OK) return error;
    execution_error = lana_effect_execute_planned(vm, plan, executor, context, out);
    ended = time(NULL); if (ended < 0) return LANA_ERR_IO;
    attempt.struct_size = sizeof(attempt); attempt.schema_version = 1u; attempt.attempt_id = attempt_id;
    attempt.decision_id = decision->decision_id; attempt.effect_desc = decision->effect;
    attempt.target = decision->target;
    attempt.status = execution_error == LANA_OK ? EFFECT_STATUS_SUCCEEDED : EFFECT_STATUS_FAILED;
    attempt.start_time = (uint64_t)started; attempt.end_time = (uint64_t)ended;
    attempt.failure_reason = execution_error == LANA_OK ? "" : lana_error_name(execution_error);
    if (out_attempt != NULL) *out_attempt = attempt;
    if ((error = store_attempt(store, vm, &attempt, true)) != LANA_OK) return error;
    if ((error = lana_store_commit(store, NULL)) != LANA_OK) return error;
    return execution_error;
}

File: runtime/rust/lana-runtime/src/effects.rs
//! Effect execution and attempt storage, mirroring `runtime/c/effects.c` and
//! `runtime/include/effects.h`.

use std::sync::Arc;
use std::time::{SystemTime, UNIX_EPOCH};

use lana_bytecode::LanaError;
use lana_vm::value::Value;

use crate::policy::{Decision, PolicyOutcome};
use crate::store::{Store, store_commit, store_get, store_put};

#[derive(Clone, Copy, PartialEq, Eq)]
#[repr(u32)]
pub enum EffectStatus {
    Pending = 0,
    Succeeded,
    Failed,
    Cancelled,
}

pub struct EffectAttempt {
    pub schema_version: u32,
    pub attempt_id: u64,
    pub decision_id: u64,
    pub effect_desc: Arc<str>,
    pub target: Option<Arc<str>>,
    pub status: EffectStatus,
    pub start_time: u64,
    pub end_time: u64,
    pub failure_reason: Arc<str>,
}

/// An effect executor, mirroring `LanaEffectExecutor`.
pub type EffectExecutor = Box<dyn Fn(&Value) -> Result<Value, LanaError>>;

fn json_safe(text: Option<&str>) -> bool {
    match text {
        None => true,
        Some(text) => text.bytes().all(|b| b >= 0x20 && b != b'\\' && b != b'"'),
    }
}

fn now_seconds() -> Result<u64, LanaError> {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_secs())
        .map_err(|_| LanaError::Io)
}

fn store_attempt(store: &mut Store, attempt: &EffectAttempt, allow_overwrite: bool) -> Result<(), LanaError> {
    if attempt.schema_version != 1
        || attempt.attempt_id == 0
        || attempt.decision_id == 0
        || attempt.effect_desc.is_empty()
        || (attempt.status as u32) > EffectStatus::Cancelled as u32
    {
        return Err(LanaError::Schema);
    }
    if !json_safe(Some(&attempt.effect_desc))
        || !json_safe(attempt.target.as_deref())
        || !json_safe(Some(&attempt.failure_reason))
    {
        return Err(LanaError::Schema);
    }
    let key = format!("effect-attempt/{}/{}", attempt.decision_id, attempt.attempt_id);
    if !allow_overwrite {
        match store_get(store, &key) {
            Ok(_) => return Err(LanaError::Conflict),
            Err(LanaError::NotFound) => {}
            Err(e) => return Err(e),
        }
    }
    let json = format!(
        "{{\"attempt_id\":{},\"decision_id\":{},\"effect\":\"{}\",\"target\":\"{}\",\"end_time\":{},\"failure_reason\":\"{}\",\"schema\":1,\"start_time\":{},\"status\":{}}}",
        attempt.attempt_id,
        attempt.decision_id,
        attempt.effect_desc,
        attempt.target.as_deref().unwrap_or(""),
        attempt.end_time,
        attempt.failure_reason,
        attempt.start_time,
        attempt.status as u32
    );
    store_put(store, &key, &Value::string(Arc::from(json.as_str())))
}

/// Execute a planned effect through the caller-supplied executor, mirroring
/// `lana_effect_execute_planned`.
pub fn effect_execute_planned(
    executor: &EffectExecutor,
    plan: &Value,
) -> Result<Value, LanaError> {
    executor(plan)
}

/// Store a PENDING attempt, mirroring `lana_effect_store_attempt`.
pub fn effect_store_attempt(store: &mut Store, attempt: &EffectAttempt) -> Result<(), LanaError> {
    store_attempt(store, attempt, false)
}

/// Execute an effect attempt end-to-end, mirroring `lana_effect_execute_attempt`.
pub fn effect_execute_attempt(
    store: &mut Store,
    decision: &Decision,
    attempt_id: u64,
    plan: &Value,
    executor: &EffectExecutor,
) -> Result<(Value, EffectAttempt), LanaError> {
    if decision.outcome != PolicyOutcome::Authorize || decision.effect.is_none() || attempt_id == 0 {
        return Err(LanaError::Capability);
    }
    let started = now_seconds()?;
    let pending = EffectAttempt {
        schema_version: 1,
        attempt_id,
        decision_id: decision.decision_id,
        effect_desc: decision.effect.clone().unwrap(),
        target: Some(decision.target.clone()),
        status: EffectStatus::Pending,
        start_time: started,
        end_time: 0,
        failure_reason: Arc::from(""),
    };
    store_attempt(store, &pending, false)?;
    store_commit(store)?;

    let execution = effect_execute_planned(executor, plan);
    let ended = now_seconds()?;
    let execution_error = execution.as_ref().err().copied();
    let attempt = EffectAttempt {
        schema_version: 1,
        attempt_id,
        decision_id: decision.decision_id,
        effect_desc: decision.effect.clone().unwrap(),
        target: Some(decision.target.clone()),
        status: if execution_error.is_none() { EffectStatus::Succeeded } else { EffectStatus::Failed },
        start_time: started,
        end_time: ended,
        failure_reason: match execution_error {
            Some(e) => Arc::from(e.name()),
            None => Arc::from(""),
        },
    };
    store_attempt(store, &attempt, true)?;
    store_commit(store)?;

    match execution {
        Ok(value) => Ok((value, attempt)),
        Err(e) => Err(e),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn store_attempt_rejects_duplicate() {
        let dir = std::env::temp_dir().join(format!("lana_effects_{}", std::process::id()));
        let _ = std::fs::remove_dir_all(&dir);
        let path = dir.to_str().unwrap().to_string();
        let mut store = crate::store::store_open(&crate::store::StoreOptions {
            schema_version: 1,
            path: path.clone(),
            timeout_ms: 0,
        })
        .unwrap();
        let attempt = EffectAttempt {
            schema_version: 1,
            attempt_id: 1,
            decision_id: 1,
            effect_desc: Arc::from("grant"),
            target: None,
            status: EffectStatus::Pending,
            start_time: 0,
            end_time: 0,
            failure_reason: Arc::from(""),
        };
        effect_store_attempt(&mut store, &attempt).unwrap();
        crate::store::store_commit(&mut store).unwrap();
        assert_eq!(effect_store_attempt(&mut store, &attempt).unwrap_err(), LanaError::Conflict);
        let _ = std::fs::remove_dir_all(&dir);
    }
}
